driver-async/src/main/com/mongodb/async/client/ClientSessionImpl.java

```diff
@@ -0,0 +1,141 @@
+        this.executor = executor;
+        if (options.getAutoStartTransaction()) {
+            startTransaction(options.getDefaultTransactionOptions());
+        }
+   }
+
+    @Override
+    public boolean hasActiveTransaction() {
+        return inTransaction;
+    }
+
+    @Override
+    public TransactionOptions getTransactionOptions() {
+        isTrue("in transaction", inTransaction);
+        return transactionOptions;
+    }
+
+    @Override
+    public void startTransaction() {
+        startTransaction(TransactionOptions.builder().build());
+    }
+
+    @Override
+    public void startTransaction(final TransactionOptions transactionOptions) {
+        notNull("transactionOptions", transactionOptions);
+        if (inTransaction) {
+            throw new IllegalStateException("Transaction already in progress");
+        }
+        inTransaction = true;
+        this.transactionOptions = TransactionOptions.merge(transactionOptions, getOptions().getDefaultTransactionOptions());
+    }
+
+    @Override
+    public void commitTransaction(final SingleResultCallback<Void> callback) {
+        if (!inTransaction) {
+            throw new IllegalStateException("There is no transaction started");
+        }
+        endTransaction(new CommitTransactionOperation(transactionOptions.getWriteConcern()), callback);
+    }
+
+    @Override
+    public void abortTransaction(final SingleResultCallback<Void> callback) {
+        if (!inTransaction) {
+            throw new IllegalStateException("There is no transaction started");
+        }
+        endTransaction(new AbortTransactionOperation(transactionOptions.getWriteConcern()), new SingleResultCallback<Void>() {
+            @Override
+            public void onResult(final Void result, final Throwable t) {
+                // Don't report failure to abort the transaction
+                callback.onResult(null, null);
+            }
+        });
+    }
+
+    private void endTransaction(final TransactionOperation operation, final SingleResultCallback<Void> callback) {
+        if (getServerSession().getStatementId() == 0) {
+            cleanupTransaction();
+            callback.onResult(null, null);
+        } else {
+            ReadConcern readConcern = transactionOptions.getReadConcern();
+            if (readConcern == null) {
+                throw new MongoInternalException("Invariant violated.  Transaction options read concern can not be null");
+            }
+            executor.execute(operation,
+                    ReadPreference.primary(), readConcern, this,
+                    new SingleResultCallback<Void>() {
+                        @Override
+                        public void onResult(final Void result, final Throwable t) {
+                            cleanupTransaction();
+                            callback.onResult(result, t);
+                        }
+                    });
+        }
+    }
+
+    // TODO: should there be a version of this that takes a callback?
```

Comment:
I started a comment thread on the spec: https://docs.google.com/a/10gen.com/document/d/14Ouqk1t7JG2hCqEQ-f7e2qgWHH5SeVpXnjybKKLtpF4/edit?disco=AAAAB3eNVtQ.  Let's leave as is until that's resolved.  Also, .NET driver did it this way too.

Reply:
This is for the `close` method rather than the `startTransaction` method. 

As starting a session takes a callback its only normal for the close to take one also for the cleanup.